File: DesignPatternsExercises/ruby/abstract_factory_design_pattern/solution/us_post_code.rb
require_relative 'post_code'

class USPostCode < PostCode
  # def __________(16)__________(post_code)
  def initialize(post_code)
    @post_code = post_code if is_valid_post_code(post_code)
    raise 'Invalid US post code' unless is_valid_post_code(post_code)

    puts "Creating instance of USPostCode: #{post_code}"
  end

  # def __________(17)__________(post_code)
  def is_valid_post_code(post_code)
    post_code =~ /^[0-9]{5}(?:-[0-9]{4})?$/
  end
end

File: DesignPatternsExercises/java/FactoryMethodDesignPattern/Exercise/USPostCode.java
public class USPostCode implements PostCode {
    private String postalCode;

    public USPostCode(String postalCode) {
        if (!__________(8)__________(postalCode)) {
            throw new IllegalArgumentException("Invalid postcode");
        }
        this.postalCode = postalCode;
        System.out.println("Creating instance of USPostCode:");
    }

    @Override
    public boolean isValidPostalCode(String postalCode) { //> 6
        return postalCode.matches("^[0-9]{5}(?:-[0-9]{4})?$");
    }

    @Override
    public String getPostalCode() {
        return postalCode;
    }
}
